core/shared/src/main/scala/scalaz/zio/internal/Executor.scala

```diff
@@ -0,0 +1,203 @@
+        cause.printStackTrace
+    }
+}
+
+object Executor extends Serializable {
+  sealed abstract class Type extends Product with Serializable
+  object Type extends Serializable {
+
+    /**
+     * An executor optimized for synchronous tasks, which yield
+     * to the runtime infrequently or never.
+     */
+    final case object Synchronous extends Type
+
+    /**
+     * An executor optimized for asynchronous tasks, which yield
+     * frequently to the runtime.
+     */
+    final case object Asynchronous extends Type
+  }
+
+  /**
+   * Creates a new default executor of the specified type.
+   */
+  final def newDefaultExecutor(tpe: Type): Executor = tpe match {
+    case Type.Synchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Int.MaxValue
+        val maxPoolSize   = Int.MaxValue
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new LinkedBlockingQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio-blocking", true)
+
+        val threadPool = new ThreadPoolExecutor(
+          corePoolSize,
+          maxPoolSize,
+          keepAliveTime,
+          timeUnit,
+          workQueue,
+          threadFactory
+        )
+        threadPool.allowCoreThreadTimeOut(true)
+
+        threadPool
+      }
+
+    case Type.Asynchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Runtime.getRuntime.availableProcessors() * 2
+        val maxPoolSize   = corePoolSize
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new LinkedBlockingQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio", true)
+
+        val threadPool = new ThreadPoolExecutor(
+          corePoolSize,
+          maxPoolSize,
+          keepAliveTime,
+          timeUnit,
+          workQueue,
+          threadFactory
+        )
+        threadPool.allowCoreThreadTimeOut(true)
+
+        threadPool
+      }
+  }
+
+  /**
+   * Creates an `Executor` from a Scala `ExecutionContext`. You must manually
+   * supply the concurrency level, since `ExecutionContext` cannot
+   * provide this detail.
+   */
+  final def fromExecutionContext(n: Int, ec: ExecutionContext): Executor =
+    new Executor {
+      import java.util.concurrent.atomic.AtomicLong
+
+      val _enqueuedCount = new AtomicLong()
+      val _dequeuedCount = new AtomicLong()
+
+      def concurrency: Int = n
+
+      def capacity: Int = Int.MaxValue
+
+      def size: Int = (enqueuedCount - dequeuedCount).toInt
+
+      def enqueuedCount: Long = _enqueuedCount.get
+
+      def dequeuedCount: Long = _dequeuedCount.get
+
+      def submit(runnable: Runnable): Boolean =
+        try {
+          ec.execute(new Runnable() {
+            def run: Unit =
+              try runnable.run()
+              finally {
+                val _ = _dequeuedCount.decrementAndGet()
```

Comment:
Should this be `incrementAndGet`? `_dequeuedCount` will grow more negative over time, that doesn't seem right.

Reply:
Good catch! 🤦‍♂️